src/shims/fs.rs

```diff
@@ -160,20 +165,24 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             return Ok(0);
         }
         let fd = this.read_scalar(fd_op)?.to_i32()?;
-        let buf_scalar = this.read_scalar(buf_op)?.not_undef()?;
-
-        // Remove the file handle to avoid borrowing issues.
-        this.remove_handle_and(fd, |mut handle, this| {
-            // Don't use `?` to avoid returning before reinserting the handle.
-            let bytes = this.force_ptr(buf_scalar).and_then(|buf| {
-                this.memory
-                    .get_mut(buf.alloc_id)?
-                    .get_bytes_mut(&*this.tcx, buf, Size::from_bytes(count))
-                    .map(|buffer| handle.file.read(buffer))
-            });
-            this.machine.file_handler.handles.insert(fd, handle).unwrap_none();
-            this.try_unwrap_io_result(bytes?.map(|bytes| bytes as i64))
-        })
+        let buf = this.read_scalar(buf_op)?.not_undef()?;
+
+        if let Some(handle) = this.machine.file_handler.handles.get_mut(&fd) {
+            // We want to read at most `count` bytes
+            let mut bytes = vec![0; count as usize];
+            let result = handle.file.read(&mut bytes).map(|c| c as i64);
+            let written_count = this.try_unwrap_io_result(result)?;
+            // `try_unwrap_io_result` returns Ok(`-1`) if `result` is an error. There is no other
+            // way of returning `-1` because the `Ok` variant of `result` contains the number of
+            // written bytes, which is a possitive value.
```

Comment:
```suggestion
            // written bytes, which is a positive value.
```
But also, what about overflow?

Reply:
I think instead of trying to reconstruct the error info from this, it is better to inspect `result`.

Reply:
what overflow am I missing? 

by inspect you mean something like?
```rust
match result {
    Ok(-1) | Err(_) => { // don't write }
    _ => { // write }
}
```

Reply:
> what overflow am I missing?

Is it possible to write more than `isize::MAX` bytes?

> I'll check if result is Ok and write, otherwise I'll pass it to try_unwrap_io_result

Yes, that's what I meant.

Reply:
> Is it possible to write more than `isize::MAX` bytes?

That would be a violation of the function contract somehow, because you're supposed to write at most `i64::MAX` bytes, but I don't see any reason to believe that someone wouldn't break it. 

However, we would discover that we wrote more than `isize::MAX` bytes after the writing operation :P. Should I do the check there?